### DISCUSSION TITLE
Model Observers Database Transaction (official documentation)

### DISCUSSION THREAD

**Discussion by discoverlance-com** [laravel/framework#45382] [Q&A] [answered]
On the laravel official documentation, under [observers and database transactions](https://laravel.com/docs/9.x/eloquent#observers-and-database-transactions), there's this note I don't understand very well: 

_When models are being created within a database transaction, you may want to instruct an observer to only execute its event handlers after the database transaction is committed. You may accomplish this by defining an $afterCommit property on the observer. **If a database transaction is not in progress, the event handlers will execute immediately**:_.

As far as the `$afterCommit` property is concerned, if I follow the explanation, it should be that the observer only executes after a commit has succeeded so I am not sure why in that last sentence it basically says that _if a transaction is not in progress, the event handlers are executed immediately_. Should the handlers not only execute after a successful transaction? Here, it says if a transaction is not in progress it will execute anyways...

Am I not understanding it clearly here or?

On second thought, I think perhaps I get the idea, the statement comes after another that says that the *the event handlers are executed after the transaction is commited* so I am assuming that the sentence about the transaction not in progress basically talking about after the transaction has been committed and is no longer pending.

If what I thought of secondly is correct, I still think it's a bit confusing to mistake it as saying it executes anytime a transaction is not in progress, perhaps if we could make it more obvious like **After the transaction has been successfully committed and is no longer in progress, the event handlers will execute immediately**. Or remove that statement altogether because it's already mentioned that the handlers execute after a transaction.

**Comment by XedinUnknown** [comment, marked as answer]
I'm new to Laravel, but not new to software engineering. The point of this hook and the transaction factor seems to be the ability to react when a record is actually been changed/created. If no transaction is in progress, this means that we are in "auto-commit" mode, which wraps every query with a transaction that is automatically committed.

So, by specifying `$afterCommit = true`,  you are probably saying in Laravel language: "I want this to run only when the change is actually persisted for all users", which happens automatically on every non-transactional query.

**Reply by discoverlance-com** [reply]
Yes, I do seem to understand it now after re-reading my question, been a while since I had this question. It makes sense, as you said, the `$afterCommit = true` should only be working when in a transaction or running your laravel database queries in a transaction. I assumed that by transaction, it meant a database operation and not literally a database transation. Thanks for the further explanation.
